# Description: This script processes raw FASTQ files for genomic analysis, including FASTQ quality control, alignment, sorting, indexing, deduplication, and QC of deduplicated BAM files.
# Author：Kui Chen → kui.chen@uhn.ca ########
# Date Created: 2021-10-01 #################
# Last Modified: 2023-12-10 ################

##########################################################################################
# Runnning at H4H cluster with ≥ 180GB memory (veryhimem) configuration ##################
# Running well with R/3.5.0 module at H4H ################################################
# Before starting, make working directory "/path/to/your/wig" ############################
##########################################################################################


.libPaths("/cluster/home/username/R/x86_64-pc-linux-gnu-library/3.5")
library(MEDIPS)
library(edgeR)
library(BSgenome.Hsapiens.UCSC.hg19)


# Define the working directory and output directory
bamDir <- "/path/to/your/bam"  #path to the BAM_DIR generated by fastq_to_bam.sh
outdir <- "/path/to/your/wig"

# Check if the output directory exists, if not, create it
if (!dir.exists(outdir)) {
  dir.create(outdir, recursive = TRUE)
}

## Global parameters
BSgenome <- "BSgenome.Hsapiens.UCSC.hg19"
uniq <- 1e-3
extend <- 300
shift <- 0
ws <- 300
paired <- TRUE
chr.select <- paste0("chr",c(1:22,"M"))

## Generate wig from bam
BamsList <- list.files(path=bamDir, pattern="\\.bam$")
print(BamsList)

SetCreated.list <- list()
for(i in 1:length(BamsList)) {

SetCreated.list [[i]] <- MEDIPS.createSet(file= BamsList[[i]],BSgenome = BSgenome, extend = extend, shift = shift, uniq = uniq, window_size = ws, chr.select = chr.select, paired = paired)

}

names(SetCreated.list) <- BamsList

for(i in 1:length(SetCreated.list)) {
 MEDIPS.exportWIG(Set=SetCreated.list[[i]], file= paste0(BamsList[[i]],".wig"), format="count", descr= BamsList[[i]])
}

