# Description: This script conducts QC of deduplicated BAM files using MEDIPS, especially for saturation & enrichment testing on IP reaction during MeDIP. The output metrics were used to generate Fig. S2A.
# Author：Kui Chen → kui.chen@uhn.ca
# 1st Created: 2020-06-21
# Last Modified: 2021-11-12

##########################################################################################
# Package version : MEDIPS v1.50.0
# Before starting, make working directory "/path/to/your/MEDIPS_QC"
# chromosome X and Y were included
##########################################################################################

library(MEDIPS)
library(BSgenome.Hsapiens.UCSC.hg19)

# Define the working directory and output directory
bamDir <- "/path/to/your/bam"  #path to the BAM_DIR generated by fastq_to_bam.sh
outdir <- "/path/to/your/MEDIPS_QC"

# Ensure output directories exist
if (!dir.exists(outdir)) dir.create(outdir, recursive = TRUE)
pdfDir <- file.path(outdir, "pdf")
if (!dir.exists(pdfDir)) dir.create(pdfDir)
qcStatsDir <- file.path(outdir, "QCstats")
if (!dir.exists(qcStatsDir)) dir.create(qcStatsDir)
pngDir <- file.path(outdir, "png")
if (!dir.exists(pngDir)) dir.create(pngDir)

# Global parameters
BSgenome="BSgenome.Hsapiens.UCSC.hg19"
uniq <- 1
extend <- 300
shift <- 0
ws <- 300
paired <- TRUE
chr.select <- paste0("chr",c(1:22,"X","Y","M"))

# Picks up deduplicated-sorted bam files
BamsList <- list.files(path=bamDir, full.names =T,pattern = "\\.bam$")
print(BamsList)

QCstats_Mat <- matrix(NA, nrow=length(BamsList), ncol=13)
for (i in 1:length(BamsList)) {
    bamFile=BamsList[[i]]
    sample_name <- basename(bamFile)

    er=MEDIPS.CpGenrich(file = bamFile, BSgenome = BSgenome, extend = extend, shift = shift, paired=paired, uniq = uniq, chr.select=chr.select)
    cr=MEDIPS.seqCoverage(file = bamFile, pattern = "CG", BSgenome = BSgenome, extend = extend, paired=paired, shift = shift, uniq = uniq, chr.select=chr.select)
    crF=MEDIPS.seqCoverage(file = bamFile, pattern = "CG", BSgenome = BSgenome, extend = extend, shift = shift, paired=paired, uniq = 0, chr.select=chr.select)
    sr=MEDIPS.saturation(file= bamFile, BSgenome = BSgenome, extend = extend, shift = shift, uniq = uniq, paired=paired, window_size = ws, nit=10, nrit=1, empty_bins=TRUE, rank=FALSE, chr.select=chr.select)

     pdf(file=paste0(outdir,"/pdf","/",sample_name,"_SaturationPlot.pdf"))
     MEDIPS.plotSaturation(sr)
     dev.off()

     pdf(file=paste0(outdir,"/pdf","/",sample_name,"_SeqCoveragePlot_Pie.pdf"))
     MEDIPS.plotSeqCoverage(seqCoverageObj = cr, type = "pie", cov.level = c(0, 1, 2, 3, 4, 5))
     dev.off()

     pdf(file=paste0(outdir,"/pdf","/",sample_name,"_SeqCoveragePlot_Hist.pdf"))
     MEDIPS.plotSeqCoverage(seqCoverageObj = cr, type = "hist", cov.level = c(0, 1, 2, 3, 4, 5), t=30)
     dev.off()

     pdf(file=paste0(outdir,"/pdf","/",sample_name,"_SeqCoveragePlot_Pie_uniqF.pdf"))
     MEDIPS.plotSeqCoverage(seqCoverageObj = crF, type = "pie", cov.level = c(0, 1, 2, 3, 4, 5))
     dev.off()

     pdf(file=paste0(outdir,"/pdf","/",sample_name,"_SeqCoveragePlot_Hist_uniqF.pdf"))
     MEDIPS.plotSeqCoverage(seqCoverageObj = crF, type = "hist", cov.level = c(0, 1, 2, 3, 4, 5), t=30)
     dev.off()

    print ("Start Coverage")
    #generating the seqCoverage just on the unique reads
    cov.level = c(0, 1, 2, 3, 4, 5)
    cov.res = cr$cov.res
    numberPos = length(cov.res)
    pattern = cr$pattern
    numberReads = cr$numberReads
    numberReadsWO = cr$numberReadsWO
    numberReadsWO_percentage = round((numberReadsWO/numberReads * 100), digits = 2)
    results = NULL
    for (j in 1:length(cov.level)) {
        if (j == 1) {
                results = c(results, length(cov.res[cov.res <= cov.level[j]]))
            }
        else {
            results = c(results, length(cov.res[cov.res > cov.level[j - 1] & cov.res <= cov.level[j]]))
            }
        }

    results = c(results, length(cov.res[cov.res > cov.level[length(cov.level)]]))
    print(results)
    print("Got the results")
    QCstats_Mat[i,1]=sample_name
    QCstats_Mat[i,2]=crF$numberReads
    QCstats_Mat[i,3]=cr$numberReads
    QCstats_Mat[i,4]=er$enrichment.score.GoGe
    QCstats_Mat[i,5]=er$enrichment.score.relH
    QCstats_Mat[i,6]=results[1]
    QCstats_Mat[i,7]=results[2]
    QCstats_Mat[i,8]=results[3]
    QCstats_Mat[i,9]=results[4]
    QCstats_Mat[i,10]=results[5]
    QCstats_Mat[i,11]=results[6]
    QCstats_Mat[i,12]=results[7]
    QCstats_Mat[i,13]=cr$numberReadsWO
}
colnames(QCstats_Mat)=c("Sample", "numReads_Uniq=F", "numReads_Uniq=T", "EnrichmentScore_GoGe", "EnrichmentScore_relH", "CpG_Seq_Coverage<=0x", "CpG_Seq_Coverage=1-1x", "CpG_Seq_Coverage=2-2x", "CpG_Seq_Coverage=3-3x", "CpG_Seq_Coverage=4-4x", "CpG_Seq_Coverage=5-5x", "CpG_Seq_Coverage=>5x", "Reads_do_not_cover_CpG" )

write.table(QCstats_Mat, file=paste0(outdir,"/QCstats","/_QCstats.txt"), col.names=T, row.names=F, quote=F, sep="\t", append=F)

for (i in 1:length(BamsList)) {
  bamFile <- BamsList[[i]]

# set sample name
sample_name <- basename(bamFile)
# run MEDIPS and extract counts
MeDIP.set <- MEDIPS.createSet(file=bamFile, BSgenome=BSgenome, extend=extend, shift=shift, uniq=uniq, window_size=ws, paired=paired, chr.select=chr.select)
# coupling
CS <- MEDIPS.couplingVector(pattern="CG", refObj=MeDIP.set)
# get saturation metrics
sr <- MEDIPS.saturation(file= bamFile, BSgenome = BSgenome, extend = extend, shift = shift, uniq = uniq, paired=paired, window_size = ws, nit=10, nrit=1, empty_bins=TRUE, rank=FALSE, chr.select=chr.select)
# Coverage
cr <- MEDIPS.seqCoverage(file=bamFile, pattern="CG", BSgenome=BSgenome, extend=extend, shift=shift, uniq=uniq, paired=paired, chr.select=chr.select)
crF <- MEDIPS.seqCoverage(file=bamFile, pattern="CG", BSgenome=BSgenome, extend=extend, shift=shift, uniq=0, paired=paired, chr.select=chr.select)
# CpG enrichment
er <- MEDIPS.CpGenrich(file=bamFile, BSgenome=BSgenome, extend=extend, shift=shift, uniq=uniq, paired=paired, chr.select=chr.select)
# write out counts
#write.table(data.frame(MeDIP.set@genome_count), paste0(outdir, "/genome_",sample_name,"_count.txt"), row.names=F, quote=F, col.names=F)

# write out saturation metrics
saturation_df <- data.frame(maxEstCorReads=sr$maxEstCor[1], maxEstCor=sr$maxEstCor[2], maxTruCorReads=sr$maxTruCor[1], maxTruCor=sr$maxTruCor[2])
rownames(saturation_df) <- sample_name
write.table(saturation_df, file=paste0(outdir, "/QCstats","/",sample_name,"_saturation.txt"), sep="\t", row.names=T, quote=F, col.names=NA)

# write out coverage metrics
coverage_df <- data.frame(nReadsCG=cr$numberReads, nReadsWOCG=cr$numberReadsWO, nReadsCG_F=crF$numberReads, nReadsWOCG_F=crF$numberReadsWO)
rownames(coverage_df) <- sample_name
write.table(coverage_df, file=paste0(outdir, "/QCstats","/",sample_name,"_coverage.txt"), sep="\t", row.names=T, quote=F, col.names=NA)

# write out enrichment metrics
enrichment_df <- t(data.frame(unlist(er)))
rownames(enrichment_df) <- sample_name
write.table(enrichment_df, file=paste0(outdir,"/QCstats","/",sample_name,"_enrichment.txt"), sep="\t", row.names=T, quote=F, col.names=NA)

# saturation & enrichment metrics were used to generate Fig. S2A ###################################

# get windows
   no.window <- NULL
   for(i in 1:length(chr.select)){
     no.window <- c(no.window, ceiling(MeDIP.set@chr_lengths[i]/ws))
   }
   window.last <- NULL
   for(i in 1:length(chr.select)){
     window.last <- c(window.last, sum(no.window[1:i]))
   }
window.first <- window.last-no.window+1
# write out window co-ordinates
write.csv(data.frame(chr=chr.select, window.first, window.last, no.window=no.window), paste0(outdir, "/MEDIPS_", sample_name,"_window_per_chr.csv"), row.names=F)

png(file=paste0(outdir,"/png","/",sample_name,"_saturation.png"), units="px", height=1000, width=1000)
MEDIPS.plotSaturation(sr)
dev.off()
# write coverage metrics
png(file=paste0(outdir,"/png","/coverage_",sample_name,".png"), units="px", height=1000, width=1000)
MEDIPS.plotSeqCoverage(seqCoverageObj=cr, type="hist", t=15, main="Sequence pattern coverage, histogram")
dev.off()
# write calibration plot
png(file=paste0(outdir,"/png","/calibration_",sample_name,".png"), units="px", height=1000, width=1000)
MEDIPS.plotCalibrationPlot(CSet=CS, main="Calibration Plot", MSet=MeDIP.set, rpkm=TRUE, xrange=TRUE)
dev.off()
}
